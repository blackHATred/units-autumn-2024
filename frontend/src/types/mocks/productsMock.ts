import { Product } from '../Product';

export const productsMock: Product[] = [
    {
        id: 1,
        name: 'IPhone 14 Pro',
        description: 'Latest iphone, buy it now',
        price: 999,
        priceSymbol: '$',
        category: 'Электроника',
        imgUrl: '/iphone.png',
    },
    {
        id: 2,
        name: 'Костюм гуся',
        description: 'Запускаем гуся, работяги',
        price: 1000,
        priceSymbol: '₽',
        category: 'Одежда',
    },
    {
        id: 3,
        name: 'Настольная лампа',
        description: 'Говорят, что ее использовали в pixar',
        price: 699,
        category: 'Для дома',
        imgUrl: '/lamp.png',
    },
    {
        id: 4,
        name: 'Принтер',
        description: 'Незаменимая вещь для студента',
        price: 7000,
        category: 'Электроника',
    },
];
