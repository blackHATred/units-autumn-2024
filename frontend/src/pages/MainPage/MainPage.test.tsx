import React from 'react';
import { render, fireEvent, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MainPage } from './MainPage';
import { useCurrentTime, useProducts } from '../../hooks';
import { productsMock } from '../../types/mocks/productsMock';
import { Category } from '../../types';

jest.mock('../../hooks');
const currentTime = '12:00:00';

describe('MainPage test', () => {
    let rendered: ReturnType<typeof render>;
    let categoryButton: HTMLElement;
    let targetCategories: Category[]; // выбранные категории, которые мы хотим увидеть
    const checkProductCards = () => {
        const productContainer = rendered.getAllByTestId('product_container');
        const filteredProducts = productsMock.filter((product) =>
            targetCategories.includes(product.category)
        );
        const otherProducts = productsMock.filter(
            (product) => !targetCategories.includes(product.category)
        );
        expect(productContainer.length).toEqual(filteredProducts.length);
        // проверяем, что отображаются только продукты из выбранных категорий
        filteredProducts.forEach((product, index) => {
            expect(
                within(productContainer[index]).getByText(product.description)
            ).toBeInTheDocument();
        });
        otherProducts.forEach((product) => {
            expect(rendered.queryByText(product.name)).not.toBeInTheDocument();
        });
    };

    beforeEach(() => {
        jest.mocked(useProducts).mockReturnValue(productsMock);
        jest.mocked(useCurrentTime).mockReturnValue(currentTime);
        rendered = render(<MainPage />);
        const categories = rendered.getByTestId('categories');
        categoryButton = within(categories).getByText(productsMock[1].category);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render correctly', () => {
        expect(rendered.asFragment()).toMatchSnapshot();
    });

    it('should display the current time', () => {
        const timeElement = rendered.getByTestId('current_time');
        expect(timeElement).toHaveTextContent(currentTime);
    });

    it('should display products of the selected category when category is clicked', () => {
        targetCategories = [productsMock[1].category];
        fireEvent.click(categoryButton);
        checkProductCards();
    });

    it('should render ProductCard for everything after unselecting all selected categories', () => {
        targetCategories = [
            productsMock[0].category,
            productsMock[1].category,
            productsMock[2].category,
        ];
        fireEvent.click(categoryButton);
        fireEvent.click(categoryButton);
        checkProductCards();
    });
});
