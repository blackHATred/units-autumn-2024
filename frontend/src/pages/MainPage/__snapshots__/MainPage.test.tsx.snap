// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage test should render correctly 1`] = `
<DocumentFragment>
  <div
    class="main-page"
  >
    <h2
      class="main-page__title"
    >
      VK Маркет
    </h2>
    <h3
      data-testid="current_time"
    >
      12:00:00
    </h3>
    <div
      class="main-page__parameters"
      data-testid="categories"
    >
      <div
        class="categories"
      >
        <div
          class="categories__badge"
        >
          Одежда
        </div>
        <div
          class="categories__badge"
        >
          Для дома
        </div>
        <div
          class="categories__badge"
        >
          Электроника
        </div>
      </div>
    </div>
    <div
      class="product-card"
      data-testid="product_container"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          IPhone 14 Pro
        </h2>
        <div
          class="product-card__description"
        >
          Latest iphone, buy it now
        </div>
        <h1
          class="product-card__price"
        >
          999 $
        </h1>
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
      <img
        alt="IPhone 14 Pro"
        class="product-card__image"
        src="/iphone.png"
      />
    </div>
    <div
      class="product-card"
      data-testid="product_container"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Костюм гуся
        </h2>
        <div
          class="product-card__description"
        >
          Запускаем гуся, работяги
        </div>
        <h1
          class="product-card__price"
        >
          1 000 ₽
        </h1>
        <div
          class="product-card__category"
        >
          Одежда
        </div>
      </div>
    </div>
    <div
      class="product-card"
      data-testid="product_container"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Настольная лампа
        </h2>
        <div
          class="product-card__description"
        >
          Говорят, что ее использовали в pixar
        </div>
        <h1
          class="product-card__price"
        >
          699 ₽
        </h1>
        <div
          class="product-card__category"
        >
          Для дома
        </div>
      </div>
      <img
        alt="Настольная лампа"
        class="product-card__image"
        src="/lamp.png"
      />
    </div>
    <div
      class="product-card"
      data-testid="product_container"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Принтер
        </h2>
        <div
          class="product-card__description"
        >
          Незаменимая вещь для студента
        </div>
        <h1
          class="product-card__price"
        >
          7 000 ₽
        </h1>
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
